File: src/agent/position_sizer.py
"""Position sizing using Kelly Criterion."""

from typing import Tuple


class PositionSizer:
    """Calculate position sizes using Kelly-inspired formula."""
    
    def __init__(self):
        """Initialize position sizer."""
        self.base_allocation = 0.50  # Start at 50% of capital
        self.confidence_scaling = 0.40  # Scale 50-90% based on conviction
        self.max_position_pct = 0.90  # Hard cap at 90%
        self.min_position_pct = 0.10  # Soft minimum at 10%
    
    def calculate_size(self, score: float) -> float:
        """
        Calculate position size as percentage of portfolio.
        
        Formula:
        position_size = min(
            base_allocation + (confidence × confidence_scaling),
            max_position_pct
        )
        
        Args:
            score: Expected return score (0-1)
        
        Returns:
            Position size as fraction of portfolio (0.10-0.90)
        """
        if not 0 <= score <= 1:
            raise ValueError(f"Score must be between 0 and 1, got {score}")
        
        position_size = self.base_allocation + (score * self.confidence_scaling)
        position_size = min(position_size, self.max_position_pct)
        position_size = max(position_size, self.min_position_pct)
        
        return position_size
    
    def kelly_formula(
        self,
        win_rate: float,
        avg_win: float,
        avg_loss: float,
    ) -> Tuple[float, float, float]:
        """
        Calculate Kelly Criterion position sizing.
        
        Kelly formula:
        edge = (win_rate × avg_win) - (loss_rate × avg_loss)
        odds = avg_win / avg_loss
        position_size = edge / odds
        
        Args:
            win_rate: Historical win rate (0-1)
            avg_win: Average win as fraction (e.g., 0.08 for 8%)
            avg_loss: Average loss as fraction (e.g., 0.04 for 4%)
        
        Returns:
            Tuple of (edge, odds, kelly_position_size)
        """
        if not (0 <= win_rate <= 1):
            raise ValueError(f"Win rate must be 0-1, got {win_rate}")
        if avg_win <= 0 or avg_loss <= 0:
            raise ValueError("Average win and loss must be positive")
        
        loss_rate = 1 - win_rate
        edge = (win_rate * avg_win) - (loss_rate * avg_loss)
        odds = avg_win / avg_loss
        
        if odds == 0:
            kelly_size = 0
        else:
            kelly_size = edge / odds
        
        # Clamp to safe range (avoid overleverage)
        kelly_size = max(0, min(kelly_size, self.max_position_pct))
        
        return edge, odds, kelly_size
    
    def calculate_shares(
        self,
        portfolio_value: float,
        position_pct: float,
        share_price: float,
    ) -> int:
        """
        Calculate number of shares to buy.
        
        Args:
            portfolio_value: Total portfolio value
            position_pct: Position size as fraction (e.g., 0.82)
            share_price: Current share price
        
        Returns:
            Number of whole shares (rounded down)
        """
        position_value = portfolio_value * position_pct
        shares = int(position_value / share_price)  # Round down
        return max(0, shares)

File: src/agent/position_sizer.ts
/**
 * Position Sizer - Kelly Formula Implementation
 * Links to spec Section 6 (Position Sizing)
 */

export interface PositionSizingResult {
  baseAllocation: number; // 50% base
  scoreAdjustment: number; // 40% of score
  finalAllocation: number; // Total (10-90% range)
  shares: number; // Calculated from allocation and price
}

export class PositionSizer {
  /**
   * Calculate position size using Kelly-inspired formula
   * Per spec Section 6: base_allocation(50%) + (score × 0.40), capped at 90%
   */
  calculatePositionSize(
    portfolioValue: number,
    currentPrice: number,
    score: number,
    minAllocation: number = 0.1,
    maxAllocation: number = 0.9
  ): PositionSizingResult {
    // Base allocation is 50% of portfolio
    const baseAllocation = 0.5;

    // Score adjustment: up to 40% additional
    const scoreAdjustment = score * 0.4;

    // Final allocation with bounds
    let finalAllocation = baseAllocation + scoreAdjustment;
    finalAllocation = Math.max(minAllocation, Math.min(maxAllocation, finalAllocation));

    // Calculate number of shares
    const allocationAmount = portfolioValue * finalAllocation;
    const shares = Math.floor(allocationAmount / currentPrice);

    return {
      baseAllocation,
      scoreAdjustment,
      finalAllocation,
      shares,
    };
  }

  /**
   * Calculate Kelly fraction for position sizing
   * Classic Kelly: f = (bp - q) / b, where:
   * - b = odds (potential return multiple)
   * - p = win probability
   * - q = loss probability (1 - p)
   */
  calculateKellyFraction(
    expectedReturn: number, // Estimated return if trade works
    winProbability: number,
    maxLoss: number = 0.15
  ): number {
    const q = 1 - winProbability;
    const b = Math.max(0.5, expectedReturn / maxLoss);

    let kellyFraction = (b * winProbability - q) / b;

    // Use half-Kelly for safety
    kellyFraction = kellyFraction / 2;

    // Bound to reasonable range
    return Math.max(0.1, Math.min(0.9, kellyFraction));
  }

  /**
   * Calculate shares accounting for cash constraints
   */
  calculateSharesWithCash(
    cash: number,
    currentPrice: number,
    maxAllocationOfCash: number = 0.9
  ): number {
    const maxSpend = cash * maxAllocationOfCash;
    return Math.floor(maxSpend / currentPrice);
  }

  /**
   * Validate position size (is it within reasonable bounds?)
   */
  isValidSize(shares: number, price: number, portfolioValue: number): boolean {
    if (shares <= 0) return false;

    const positionValue = shares * price;
    const allocation = positionValue / portfolioValue;

    // Check if within acceptable range
    return allocation >= 0.1 && allocation <= 0.9;
  }
}
